pkg/redirect/redirect.go

```diff
@@ -0,0 +1,324 @@
+		"url", req.URL.String(),
+		"header", redact(req.Header))
+	w.Header().Set("X-Redirected", req.URL.String())
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		logger.Errorf("Error sending request: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Infow("got response",
+		"method", req.Method,
+		"url", req.URL.String(),
+		"status", resp.Status,
+		"header", redact(resp.Header))
+
+	for k, v := range resp.Header {
+		for _, vv := range v {
+			w.Header().Add(k, vv)
+		}
+	}
+	w.WriteHeader(resp.StatusCode)
+	if _, err := io.Copy(w, resp.Body); err != nil {
+		logger.Errorf("Error copying response body: %v", err)
+	}
+}
+
+func (rdr redirect) proxy(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.FromContext(ctx)
+
+	var url string
+	if rdr.host == "gcr.io" {
+		url = "https://gcr.io/v2/"
+	} else {
+		url = "https://ghcr.io/v2/"
+	}
+	if rdr.repo != "" {
+		url += rdr.repo + "/"
+	}
+
+	path := strings.TrimPrefix(r.URL.Path, "/v2/")
+	if rdr.prefix != "" {
+		// Trim the prefix, if any.
+		// If the path didn't have the prefix, that's fine for now too.
+		path = strings.TrimPrefix(path, rdr.prefix+"/")
+	}
+
+	url += path
+	if query := r.URL.Query().Encode(); query != "" {
+		url += "?" + query
+	}
+	req, _ := http.NewRequest(r.Method, url, nil)
+	req.Header = r.Header.Clone()
+
+	// If the request is coming in without auth, get some auth.
+	// This is useful for testing, but should never happen in real life.
+	// Actually, containerd seems to make unauthenticated HEAD requests before
+	// hitting /v2/, so this might be load-bearing.
+	if req.Header.Get("Authorization") == "" {
+		t, resp, err := rdr.getToken(r)
+		if err != nil {
+			if resp != nil {
+				logger.Infof("Error response getting token: %d %s", resp.StatusCode, resp.Status)
+				http.Error(w, resp.Status, resp.StatusCode)
+				return
+			}
+			logger.Errorf("Error getting token: %v", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		req.Header.Set("Authorization", "Bearer "+t)
+	}
+
+	logger.Infow("sending request",
+		"method", req.Method,
+		"url", req.URL.String(),
+		"header", redact(req.Header))
+	w.Header().Set("X-Redirected", req.URL.String())
+
+	resp, err := http.DefaultTransport.RoundTrip(req) // Transport doesn't follow redirects.
+	if err != nil {
+		logger.Errorf("Error sending request: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Infow("got response",
+		"method", r.Method,
+		"url", r.URL.String(),
+		"status", resp.Status,
+		"header", redact(resp.Header))
+
+	for k, v := range resp.Header {
+		for _, vv := range v {
+			if k == "Link" && strings.HasPrefix(vv, "</v2/"+rdr.repo) {
+				log.Println("=== BEFORE: Link:", vv)
+				vv = "</v2" + strings.TrimPrefix(vv, "</v2/"+rdr.repo)
+				log.Println("=== CHANGED: Link:", vv)
```

Comment:
maybe a comment here on what this header looks like... having trouble reading this

Reply:
Excellent idea! Doing this made me realize this wasn't handling the new prefix logic correctly, so I fixed that too. 😎